Adds mongo read preferences for use when a replica set is provided

pomegranited opened the pull request:
Allows for different mongo `read_preference` settings to be used for the LMS, Preview, and Studio, and ensures that the mongo replicaset setting is included in all modulestore and contentstore configuration blocks.

The default value of `EDXAPP_MONGO_REPLICA_SET: ''` means no replicaset will be used.

If `EDXAPP_MONGO_REPLICA_SET` is set, then the replica set will be used for modulestore and contentstore connections for the LMS and Studio.

When a replica set is configured, the `read_preference` setting has a different default for each system:
* CMS: `PRIMARY`
  Shared by CMS content, module, and doc stores.
  Can be changed using `EDXAPP_MONGO_CMS_READ_PREFERENCE`
* LMS draft docstore: `PRIMARY`
  Can be changed using `EDXAPP_MONGO_CMS_READ_PREFERENCE`, or independently from the CMS using `EDXAPP_LMS_SPLIT_DOC_STORE_READ_PREFERENCE`.
  This provides instructors with consistent content between Preview and Studio views.
* LMS: `SECONDARY_PREFERRED`.
  Shared by LMS content, module, and doc stores (except for the draft doc store, noted above).
  Can be changed using `EDXAPP_MONGO_LMS_READ_PREFERENCE`.

**JIRA tickets**: [OSPR-1989](https://openedx.atlassian.net/browse/OSPR-1989), OC-3200, OC-2840

**Sandbox URL**:

Configured using the default, no replica set:
* LMS: https://oc3200-2.sandbox.stage.opencraft.hosting/
* Studio: https://studio-oc3200-2.sandbox.stage.opencraft.hosting/

Configured using a replica set:
* LMS: https://oc3200.sandbox.opencraft.hosting/
* Studio: https://studio-oc3200.sandbox.opencraft.hosting/

**Merge deadline**: None

**Testing instructions**:

To verify the default, no replica setup
1. Provision a new appserver without setting `EDXAPP_MONGO_REPLICA_SET`  (e.g. [oc3200-2.sandbox.stage.opencraft.hosting](https://oc3200-2.sandbox.stage.opencraft.hosting/)).
1. Verify that you can create and edit courses as usual.
1. From the LMS shell, run these commands to verify the replica set is not enabled:
   ```python
   >>> from xmodule.modulestore.django import modulestore
   >>> modulestore().default_modulestore.db_connection.database.client._rs_client
   False
   >>> modulestore().contentstore.fs_files.database.client._rs_client
   False
   ```

To verify that using this new variable configures the replica set:
1. Provision a new appserver and set `EDXAPP_MONGO_REPLICA_SET`, e.g. [oc3200.sandbox.opencraft.hosting]( https://oc3200.sandbox.opencraft.hosting/):
   ```yaml
   # Note that the MONGO_HOSTS must be in a comma-delimited string, not a list
   EDXAPP_MONGO_HOSTS: "ds231255-a0.mlab.com,ds231255-a1.mlab.com"
   EDXAPP_MONGO_REPLICA_SET: 'rs-ds231255'
   ```
1. Verify that you can create and edit courses as usual.
1. From the LMS shell, run these commands to verify the replica set is enabled:
   ```python
   >>> from xmodule.modulestore.django import modulestore
   >>> modulestore().default_modulestore.db_connection.database.client._rs_client
   True
   >>> modulestore().contentstore.fs_files.database.client._rs_client
   True
   ```

**Reviewers**
- [x] @UmanShahzad 
- [ ] @jdmulloy CC @edx/devops 

Configuration Pull Request
---

Make sure that the following steps are done before merging:

  - [ ] A DevOps team member has approved the PR.
  - [ ] Are you adding any new default values that need to be overridden when this change goes live? If so:
    - [ ] Update the appropriate internal repo (be sure to update for all our environments)
    - [ ] If you are updating a secure value rather than an internal one, file a DEVOPS ticket with details.
    - [x] Add an entry to the CHANGELOG.
  - [ ] If you are making a complicated change, have you performed the proper testing specified on the [Ops Ansible Testing Checklist](https://openedx.atlassian.net/wiki/display/EdxOps/Ops+Ansible+Testing+Checklist)?  Adding a new variable does not require the full list (although testing on a sandbox is a great idea to ensure it links with your downstream code changes).

edx-webhook commented:
Thanks for the pull request, @pomegranited! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?number=4182&repo=edx%2Fconfiguration"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

edx-webhook commented:
Thanks for the pull request, @pomegranited! I've created [OSPR-1989](https://openedx.atlassian.net/browse/OSPR-1989) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

nedbat commented:
@edx/devops Please review.

feanil commented on playbooks/roles/edxapp/defaults/main.yml:
```diff
@@ -65,12 +65,19 @@ EDXAPP_XQUEUE_DJANGO_AUTH:
   password: 'password'
 EDXAPP_XQUEUE_URL: 'http://localhost:18040'
 
+# if EDXAPP_MONGO_REPLICA_SET is set, EDXAPP_MONGO_HOSTS must use a comma-delimited string instead of a list
```
Why is this?  Dose the underlying python not handle a list well if a replica set is provided?

pomegranited replied:
Unfortunately, yes, the `MongoReplicaSetClient` expects a string., though I believe this has been fixed in later versions of the package.   Here's an example traceback: 
```
Oct 30 06:36:27 [service_variant=lms][gunicorn.error][env:sandbox] ERROR [glogging.py:219] - Exception in worker process:
Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/gunicorn/arbiter.py", line 473, in spawn_worker
    worker.init_process()
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/gunicorn/workers/base.py", line 100, in init_process
    self.wsgi = self.app.wsgi()
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/gunicorn/app/base.py", line 106, in wsgi
    self.callable = self.load()
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/gunicorn/app/wsgiapp.py", line 27, in load
    return util.import_app(self.app_uri)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/gunicorn/util.py", line 353, in import_app
    __import__(module)
  File "/edx/app/edxapp/edx-platform/lms/wsgi.py", line 32, in <module>
    modulestore()
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/modulestore/django.py", line 331, in modulestore
    contentstore(),
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/contentstore/django.py", line 28, in contentstore
    _CONTENTSTORE[name] = class_(**options)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/contentstore/mongo.py", line 42, in __init__
    port=port, tz_aware=tz_aware, user=user, password=password, proxy=proxy, **kwargs
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/mongo_utils.py", line 42, in connect_to_mongodb
    **kwargs
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/pymongo/mongo_replica_set_client.py", line 613, in __init__
    self.__seeds.update(uri_parser.split_hosts(host, port))
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/pymongo/uri_parser.py", line 229, in split_hosts
    for entity in hosts.split(','):
AttributeError: 'list' object has no attribute 'split'
```

pomegranited replied:
@feanil Can you see a better way of handling this issue?

pomegranited commented:
Rebased c553c44 on master to resolve conflicts.

feanil commented:
@pomegranited trying to understand the need for this, we currently have a replset for our clusters but don't need to provide the name of the repl set for things to work correctly.  What's this trying to fix?

Also, if we were to merge this, you're saying that an empty string is considered an invalid name for a replset and things work the same way as if the setting had not been set at all?  I just want to make sure we don't do anything weird because now the `replicaSet` key will always be present.

pomegranited commented:
@feanil Our reason for requesting this change was to enable the use of replica sets for our clients that use clustered mlab.com mongo databases.  mlab did maintenance recently which caused outages for these services. We had thought that it was sufficient to have both of the hosts in cluster in the `EDXAPP_MONGO_HOSTS` list, but this turned out not to be the case.  Because the pymongo client wasn't connecting using the replica set, failover to the second host didn't happen.  We did a edxapp service restart after the maintenance, and all came good, but we'd like to avoid outages like this in the future.

I used the name `EDXAPP_MONGO_REPLICA_SET` because of [this 27/7/2016 Slack#ops comment](https://openedx.slack.com/messages/C08B4LZEZ/) from @jibsheet.  Though now that I look at edx-private config, I can see [`EDXAPP_MONGO_REPLICA_SET` mentioned](https://github.com/edx/edx-internal/search?utf8=%E2%9C%93&q=EDXAPP_MONGO_REPLICA_SET&type=), but it's not referenced anywhere else, so maybe you're doing something different now?  I'm open to any suggestions.

This change provides a way to pass the `replicaSet` option for the contentstore and modulestore mongo connections, which causes [`xmodule.mongo_utils` to create a `pymongo.MongoReplicaSetClient`](https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/mongo_utils.py#L23-L31), instead of the usual `pymongo.MongoClient`.

We've tested using the default empty string `EDXAPP_MONGO_REPLICA_SET` as described in the PR description, and contentstore and modulestore read/write worked fine.  

Is there any more testing we can do, or do you need more information?

feanil commented:
@pomegranited we're hoping to test this tomorrow and get you a full review soon.  @jdmulloy has a similar PR to fix some read preference settings but I think your approach is a more generic version of that change.  I think I now fully understand what's going on with this, the EDXAPP_MONGO_REPLICA_SET was previously used in the overrides that were removed a little while ago from configuration but I think it makes sense to have the settings as you have theme here.

jdmulloy commented on playbooks/roles/edxapp/defaults/main.yml:
```diff
@@ -65,12 +65,19 @@ EDXAPP_XQUEUE_DJANGO_AUTH:
+  replicaSet: "{{ EDXAPP_MONGO_REPLICA_SET }}"
+EDXAPP_CONTENTSTORE_ADDITIONAL_OPTS:
+  default: *edxapp_mongo_additional_options
+  split: *edxapp_mongo_additional_options
```
Can you explain the use of the split key here? I was unable to find it the edx-platform code for contentstore and couldn't find it but maybe I missed something?

I don't believe the contentstore has a "split" store, however there is a trashcan store that should be covered. I'm still unfamiliar with the workings of the content and module store systems so I may have missed something?

pomegranited replied:
@jdmulloy Sorry for the delayed reply here!

I'm actually not that familiar with the different types of mongo contenstores available.  I can see that the [`default` one is used by default in edx-platform](https://github.com/edx/edx-platform/blob/6d2468d0dcf782a7561ecc61946146d6fae2762f/common/lib/xmodule/xmodule/contentstore/django.py#L20-L28), and I've seen examples of split modulestores used by some of our clients (like what is [tested here](https://github.com/edx/edx-platform/blob/2d19136bebc56dc623e163f2fa56fa1644b6457e/common/lib/xmodule/xmodule/modulestore/tests/test_split_modulestore.py#L65-L69)).

@mtyaka Can you shed some light here?

mtyaka replied:
@pomegranited I think "split" is only used for modulestore, not contentstore. Doing a quick search through edx-platform, it looks like the only contentstore used in the repository other than "default" is "thrashcan", for example here: https://github.com/edx/edx-platform/blob/6e7206503297a7a3ec239f689024690a0e4db681/cms/djangoapps/contentstore/views/assets.py#L516.

pomegranited replied:
Thanks @mtyaka -- so should I be referencing `trashcan` here instead of `split`?

mtyaka replied:
@pomegranited Yes, I think so.

pomegranited replied:
Ok, da73d62 does this, but I'll need to spin up some new servers to test it.

pomegranited commented:
Rebased a99e96e on master to resolve conflicts, specifically those related to https://github.com/edx/configuration/pull/4207.

jdmulloy commented:
@pomegranited This looks good. I did some testing on it today. There is one thing we're concerned about. We're concerned about the old xmodule.modulestore.mongo.DraftMongoModuleStore module having a race condition with the SECONDARY_PREFERRED setting. Can we have the read_preference value of EDXAPP_LMS_DRAFT_DOC_STORE_CONFIG be set to "PRIMARY" or not include the read preference setting? Adding the read_preference setting below `<<: *edxapp_generic_default_docstore` in EDXAPP_LMS_DRAFT_DOC_STORE_CONFIG should allow overriding the value in `*edxapp_mongo_additional_options`.

Thanks

pomegranited commented:
@jdmulloy Thank you for your feedback and testing!  I'd be interested to hear how you test for race conditions with the draft docstore, and why they're more likely to happen there than with the other stores?

dbf5842 replaces the previous `EDXAPP_LMS_SPLIT_DOC_STORE_READ_PREFERENCE` with the more general `EDXAPP_MONGO_READ_PREFERENCE`, to indicate that this is the default read preference for all docstores and contentstores.  Default is still `SECONDARY_PREFERRED`.

It then adds `EDXAPP_LMS_DRAFT_DOC_STORE_READ_PREFERENCE: "PRIMARY"` to indicate that this docstore behaves differently from the others.

If you'd rather I added individual `EDXAPP_*_READ_PREFERENCE` variables for each store, I can do that instead.

jdmulloy commented:
Thanks for the changes @pomegranited. I don't have any testing for the race conditions, it's more of a precaution that others on the team with more experience with the old modulestore and time at edx brought up. We know it generally has issues if multiple users try to edit the same course at the same time. I think the concern is that if Studio/CMS reads from a secondary and gets stale data and someone then edits that data, when Studio/CMS writes the data back it will overwrite the previous changes.

Thinking about it I'm not sure if CMS should have SECONDARY_PREFERRED in general. I'll discuss it with the team when we get in tomorrow.

pomegranited commented:
@jdmulloy That's a good point about the CMS -- I agree that the `SECODARY_PREFERRED` should  be used to improve performance for learners in the LMS.  Hits to the draft docstore are likely to be low volume, to test changes made in the CMS, but they need to be accurate, not delayed by replication.

Thanks for discussing it with your team.  I too am relying on the wisdom of more experienced team members -- @mtyaka do you have any feedback on this approach?

mtyaka commented:
@pomegranited @jdmulloy Good point about the CMS. I guess `PRIMARY` would be a safer default for all modulestore operations in the Studio. I don't have much experience with Mongo read modes though, so am curious what @jdmulloy's team thinks about it.

jdmulloy commented:
@pomegranited I talked with the team and got an explanation from @feanil about our concerns about old Draft Mongo. Our main concern is that it's brittle and unmaintained, so we don't want to surprise anyone by making changes to it's configuration. I think between the changes you already made to accommodate this and splitting the LMS and CMS settings we should be able to merge this soon.

Thanks

pomegranited commented on playbooks/roles/edxapp/defaults/main.yml:
```diff
@@ -67,14 +67,21 @@ EDXAPP_XQUEUE_URL: 'http://localhost:18040'
 EDXAPP_MONGO_REPLICA_SET: ''
-EDXAPP_LMS_SPLIT_DOC_STORE_READ_PREFERENCE: 'SECONDARY_PREFERRED'
+EDXAPP_MONGO_READ_PREFERENCE: 'SECONDARY_PREFERRED'
+EDXAPP_LMS_DRAFT_DOC_STORE_READ_PREFERENCE: 'PRIMARY'
```
@jdmulloy Should we use `PRIMARY_PREFERRED` instead here, just in case the primary is down?

pomegranited commented on playbooks/roles/edxapp/defaults/main.yml:
```diff
@@ -74,8 +74,10 @@ EDXAPP_MONGO_USER: 'edxapp'
 EDXAPP_MONGO_READ_PREFERENCE: 'SECONDARY_PREFERRED'
-EDXAPP_LMS_DRAFT_DOC_STORE_READ_PREFERENCE: 'PRIMARY'
+EDXAPP_MONGO_LMS_READ_PREFERENCE: '{{ EDXAPP_MONGO_READ_PREFERENCE }}'
+EDXAPP_MONGO_CMS_READ_PREFERENCE: 'PRIMARY'
```
@jdmulloy  I've removed the `EDXAPP_LMS_DRAFT_DOC_STORE_READ_PREFERENCE` I added, and instead am using `EDXAPP_MONGO_CMS_READ_PREFERENCE` below when specifying the [`EDXAPP_LMS_DRAFT_DOC_STORE_CONFIG `](https://github.com/edx/configuration/pull/4182/commits/d2ba154ffb9310ecdc918f92ebb0dcbc9062ff54#diff-9bffb77ca37dcc0aead699b822b29095R881).

Is this clear enough?

pomegranited commented:
Excellent, thank you @jdmulloy !  This is now ready for a (hopefully final) review.

jdmulloy commented on playbooks/roles/edxapp/defaults/main.yml:
```diff
@@ -67,14 +67,24 @@ EDXAPP_XQUEUE_URL: 'http://localhost:18040'
 EDXAPP_MONGO_USE_SSL: False
 EDXAPP_MONGO_REPLICA_SET: ''
-EDXAPP_LMS_SPLIT_DOC_STORE_READ_PREFERENCE: 'SECONDARY_PREFERRED'
+# Used iff EDXAPP_MONGO_REPLICA_SET is provided.
```
Should `iff` be `if`?

pomegranited replied:
Sure, fixed with https://github.com/edx/configuration/pull/4182/commits/cf33d5a53d9cde5c43f6b3ef09afb52b7c2a2043#diff-9bffb77ca37dcc0aead699b822b29095R77

jdmulloy commented on playbooks/roles/edxapp/defaults/main.yml:
```diff
@@ -863,18 +872,22 @@ edxapp_generic_doc_store_config: &edxapp_generic_default_docstore
   connectTimeoutMS: 2000 # default is 20000, I believe raises pymongo.errors.ConnectionFailure
   # Not setting waitQueueTimeoutMS and waitQueueMultiple since pymongo defaults to nobody being allowed to wait
-
+  <<: *edxapp_mongo_additional_options
```
This is pulling in the `PRIMARY_PREFERED` setting for CMS Docstore and Contentstore. I made a branch with a commit that fixes this. It may or may not be a good solution to the problem. See PR https://github.com/open-craft/configuration/pull/30

pomegranited replied:
@jdmulloy Good call!

cf33d5a adapts your change it to remove duplication, could you have a look?

I'm deploying some servers to test this now, and will let you know how it goes.

pomegranited commented on playbooks/roles/edxapp/defaults/main.yml:
```diff
@@ -67,14 +67,24 @@ EDXAPP_XQUEUE_URL: 'http://localhost:18040'
 # EDXAPP_MONGO_HOSTS must be a comma seperated list of hosts/ips for
 # compatibility with pymongo.MongoReplicaSetClient in PyMongo 2.9.1
-EDXAPP_MONGO_HOSTS: 'localhost'
+EDXAPP_MONGO_HOSTS: ['localhost']
```
@jdmulloy This is complicated, actually, and we should probably fix it in edx-platform instead of using just a comment here. 

Basically, `MongoClient` expects the `host` parameter to be a list of host names, and `MongoReplicaSetClient` expects the `host` parameter to be a comma-delimited string of host names.

Backwards and broken, I know, but the `host` parameter is deprecated, and all of this is changed and sane in the latest PyMongo anyway. Please see [the discussion above](https://github.com/edx/configuration/pull/4182#discussion_r148009086).

pomegranited replied:
Done with c2e8379d22f5661fff2890c59a9cf962ec009f0c, thanks for doing that testing!

pomegranited commented:
@jdmulloy CC @feanil  Argh.. this is doing my head in.  Can you help me figure out what all the pieces are, and what end result we're aiming for?

Here's what I think I know so far:

* Each system (LMS, CMS) has a module store, a content store, and a doc store.
* The module store comprises the `split` and `draft` doc stores.
* The content store comprises the `default` and `trashcan` stores.
* For legacy reasons I don't understand, the doc store config is duplicated in 6 places in the auth config for each system.  **Can you help explain why?**
  * At the [top level](https://github.com/edx/configuration/blob/2fbd8e801fd1dab8071ef419784b1d6926b0962b/playbooks/roles/edxapp/defaults/main.yml#L936), shared by both LMS and CMS.
  * Within the [CONTENTSTORE](https://github.com/edx/configuration/blob/2fbd8e801fd1dab8071ef419784b1d6926b0962b/playbooks/roles/edxapp/defaults/main.yml#L955), shared by both LMS and CMS.
  * In LMS [`split`](https://github.com/edx/configuration/blob/2fbd8e801fd1dab8071ef419784b1d6926b0962b/playbooks/roles/edxapp/defaults/main.yml#L1192) and [`draft`]https://github.com/edx/configuration/blob/2fbd8e801fd1dab8071ef419784b1d6926b0962b/playbooks/roles/edxapp/defaults/main.yml#L1199 module stores
  * In CMS [`split`](https://github.com/edx/configuration/blob/2fbd8e801fd1dab8071ef419784b1d6926b0962b/playbooks/roles/edxapp/defaults/main.yml#L1275) and [`draft`](https://github.com/edx/configuration/blob/2fbd8e801fd1dab8071ef419784b1d6926b0962b/playbooks/roles/edxapp/defaults/main.yml#L1282) module stores
* When a replica set is provided, the read preference default for the mongo client is `PRIMARY`.
* LMS should use `SECONDARY_PREFERRED` for all the learner-facing mongo stores in LMS.
* LMS DRAFT docstore should use `PRIMARY`, because it's used by instructors to preview content edited in the CMS.
* CMS should use `PRIMARY` for all mongo stores.
  However, we should keep the CMS and LMS DRAFT read_preference variables (e.g. if we want to use `PRIMARY_PREFERRED` instead someday.)

So it's looking like we need separate stanzas for all 7 of these combinations:
* CMS `split` and `draft` doc stores: `PRIMARY`
* CMS `default` and `trashcan` content stores: `PRIMARY`
* LMS `draft` doc store: `PRIMARY`
* LMS `split` doc store: `SECONDARY_PREFERRED`
* LMS `default` and `trashcan` content stores: `SECONDARY_PREFERRED`

So to reduce confusion and sources of error, 8e9df5609 moves the `CONTENTSTORE` and `DOC_STORE_CONFIG` stanzas out of the shared generic stanza, and down into the individual `lms_auth_config` and `cms_auth_config` stanzas, like what is already done with the `MODULESTORE` config.

* Added 4 variables only:
   ```yaml
   EDXAPP_MONGO_CMS_READ_PREFERENCE: 'PRIMARY'`
   EDXAPP_MONGO_LMS_READ_PREFERENCE: 'SECONDARY_PREFERRED'`
   EDXAPP_LMS_SPLIT_DOC_STORE_READ_PREFERENCE: '{{ 
   EDXAPP_MONGO_LMS_READ_PREFERENCE }}'
   EDXAPP_LMS_DRAFT_DOC_STORE_READ_PREFERENCE: '{{ 
   EDXAPP_MONGO_CMS_READ_PREFERENCE }}'
   ```
* Pulled the common contentstore config elements out into `edxapp_generic_contentstore_config`
* Moved the shared top-level `DOC_STORE_CONFIG` and `CONTENTSTORE` stanzas out of `edxapp_generic_doc_store_config`, and into the explicit definitions in `lms_auth_config` and `cms_auth_config`.
* Removed `EDXAPP_CONTENTSTORE_ADDITIONAL_OPTS`.
  This might be a problem?  But I couldn't figure out a nice way to incorporate common additional options into all the separate system contentstores that needed to be configured, without creating separate variables with merge keys, which would break existing variable configurations anyway.

jdmulloy commented:
Hi @pomegranited I had a chance to review this today and do some testing. There's a few changes that I think should be made. I made a branch that merges this branch with master and I made some minor changes to get it to generate the config we need for edx.org. I still need to do a little more testing, but I think we're really close. You can see my changes here: https://github.com/edx/configuration/commit/207c5e0f0085cb740b7547d7f0d11f0f59f2fd9c

pomegranited commented on playbooks/roles/edxapp/defaults/main.yml:
```diff
@@ -851,6 +854,20 @@ edxapp_environment_extra: {}
+  #
+  OPTIONS:
+    db: "{{ EDXAPP_MONGO_DB_NAME }}"
+    host: "{{ EDXAPP_MONGO_HOSTS }}"
```
@jdmulloy From looking at [`mongo_utils.connect_to_mongodb()`](https://github.com/edx/edx-platform/blob/622ee7a7e35e33b198561e525bafd93e61aa82c8/common/lib/xmodule/xmodule/mongo_utils.py#L27-L32), it looks like we can specify the `replicaSet` variable as an empty string, and still fall back to using `pymongo.MongoClient` instead of `pymongo.MongoReplicaSetClient`. So we don't have to have a replica set configured on the mongo service, which is nice -- e.g. for devstacks and self-hosted mongo instances.

pomegranited commented on playbooks/roles/edxapp/defaults/main.yml:
```diff
@@ -1166,6 +1170,17 @@ generic_env_config:  &edxapp_generic_env
   <<: *edxapp_generic_auth
+  CONTENTSTORE:
+    <<: *edxapp_generic_default_contentstore
+    ADDITIONAL_OPTIONS:
```
+1

pomegranited commented:
I've cherry-picked 207c5e0, and c2e8379d22f5661fff2890c59a9cf962ec009f0c changes `EDXAPP_MONGO_HOSTS` back to `localhost` as you requested.  Glad to know that works for `MongoReplicaSetClient` too!

I'm deploying a couple of new sandbox servers to verify these changes too.

Thank you for all your help!

pomegranited commented:
@jdmulloy My test servers configured as expected with the latest change.
Is there anything else you can think of that needs to be fixed or tested before we can squash and merge?

pomegranited commented:
@jdmulloy Squashed and ready to merge!  Thank you for all your help with this.

edx-webhook commented:
@pomegranited 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/configuration&entry.867055334=https://github.com/edx/configuration/pull/4182&entry.1484655318=https://github.com/pomegranited&entry.752974735=2017-10-30+06:20&entry.1917517419=2017-12-20+21:09&entry.2133058324=Yes) so we can improve your experience in the future.

coryleeio commented:
Thank you very much for your contribution @pomegranited =] 

I've merged it

pomegranited commented:
OMG YAY!  Thank you for all your help @jdmulloy @feanil @coryleeio !
